From 5a5b0f04484e00c88e7be902101367d6d591fb96 Mon Sep 17 00:00:00 2001
From: Jugurtha BELKALEM <jugurtha.belkalem@smile.fr>
Date: Thu, 2 May 2019 11:06:23 +0200
Subject: [PATCH] cmake/luajit: Provide the target architecture to luaJIT while
 cross-compiling

Unlike CMAKE_SYSTEM_PROCESSOR which identifies aarch64
as a valid architecture, luajit does not recognize it.
luajit defines aarch64 as arm64.

LuaJIT doesn't use usual arch naming, so we have to convert
between CMake and Luajit for each architectures while
cross-compiling.

Signed-off-by: Jugurtha BELKALEM <jugurtha.belkalem@smile.fr>
Signed-off-by: Romain Naour <romain.naour@smile.fr>
---
v2: Do the same for other architecture supported by LuaJIT.
https://github.com/iovisor/bcc/pull/2480
---
 src/lua/CMakeLists.txt | 29 ++++++++++++++++++++++++++++-
 1 file changed, 28 insertions(+), 1 deletion(-)

diff --git a/src/lua/CMakeLists.txt b/src/lua/CMakeLists.txt
index 7541d48df..226e1b1d2 100644
--- a/src/lua/CMakeLists.txt
+++ b/src/lua/CMakeLists.txt
@@ -13,9 +13,36 @@ if (LUAJIT_LIBRARIES AND LUAJIT)
 		DEPENDS ${SRC_LUA} ${CMAKE_CURRENT_SOURCE_DIR}/squishy
 	)
 
+	# LuaJIT doesn't use usual arch naming, so we have to convert
+	# between CMake and Luajit while cross-compiling.
+	if (CMAKE_CROSSCOMPILING)
+		SET (LUAJIT_TARGET_ARCH "-a")
+		# https://github.com/LuaJIT/LuaJIT/blob/f0e865dd4861520258299d0f2a56491bd9d602e1/src/jit/bcsave.lua#L30
+		# https://github.com/LuaJIT/LuaJIT/blob/f0e865dd4861520258299d0f2a56491bd9d602e1/src/jit/bcsave.lua#L65
+		if (CMAKE_SYSTEM_PROCESSOR STREQUAL "aarch64")
+			SET (LUAJIT_TARGET_ARCH ${LUAJIT_TARGET_ARCH} "arm64")
+		elseif (CMAKE_SYSTEM_PROCESSOR MATCHES "aarch64_be")
+			SET (LUAJIT_TARGET_ARCH ${LUAJIT_TARGET_ARCH} "arm64be")
+		elseif (CMAKE_SYSTEM_PROCESSOR MATCHES "arm")
+			SET (LUAJIT_TARGET_ARCH ${LUAJIT_TARGET_ARCH} "arm")
+		elseif (CMAKE_SYSTEM_PROCESSOR MATCHES "^(i.86)$")
+			SET (LUAJIT_TARGET_ARCH ${LUAJIT_TARGET_ARCH} "x86")
+		elseif (CMAKE_SYSTEM_PROCESSOR MATCHES "mips")
+			SET (LUAJIT_TARGET_ARCH ${LUAJIT_TARGET_ARCH} "mips")
+		elseif (CMAKE_SYSTEM_PROCESSOR MATCHES "mipsel")
+			SET (LUAJIT_TARGET_ARCH ${LUAJIT_TARGET_ARCH} "mipsel")
+		elseif (CMAKE_SYSTEM_PROCESSOR MATCHES "powerpc")
+			SET (LUAJIT_TARGET_ARCH ${LUAJIT_TARGET_ARCH} "ppc")
+		elseif (CMAKE_SYSTEM_PROCESSOR MATCHES "x86_64")
+			SET (LUAJIT_TARGET_ARCH ${LUAJIT_TARGET_ARCH} "x64")
+		else()
+			MESSAGE(FATAL_ERROR "${CMAKE_SYSTEM_PROCESSOR} is not supported by LuaJIT")
+		endif()
+	endif()
+
 	ADD_CUSTOM_COMMAND(
 		OUTPUT bcc.o
-		COMMAND ${LUAJIT} -bg bcc.lua bcc.o
+		COMMAND ${LUAJIT} -bg bcc.lua ${LUAJIT_TARGET_ARCH} bcc.o
 		DEPENDS bcc.lua
 	)
 
